Newsletters: add 112 (2020-08-26) (#454)
bitcoinops/bitcoinops.github.io: _posts/en/newsletters/2020-08-26-newsletter.md

---
title: 'Bitcoin Optech Newsletter #112'
permalink: /en/newsletters/2020/08/26/
name: 2020-08-26-newsletter
slug: 2020-08-26-newsletter
type: newsletter
layout: newsletter
lang: en
---
This week's newsletter links to a discussion about routed coinswaps and
includes our regular sections with summaries of questions and answers
from the Bitcoin StackExchange, releases and release candidates, and

> notable changes to popular Bitcoin infrastructure software

Reviewer: missing full stop here and line 30.

## Action items

*None this week.*

## News

- **Discussion about routed coinswaps:** Chris Belcher [posted][belcher
  coinswap] a design document for a routed multi-transaction coinswap
  implementation to the Bitcoin-Dev mailing list as a follow-up to his
  previous post in May (see [Newsletter #100][news100 coinswap]).
  Discussion from ZmnSCPxj and Nadav Kohen focused on ensuring the

>   protocol was safe both in its use of cryptography and in ensuring that

Reviewer 2: Remove repeated ensuring: "ensuring the protocol was safe in ... ensuring that ..."

FIXME:bitschmidty

## Releases and release candidates

*New releases and release candidates for popular Bitcoin infrastructure
projects.  Please consider upgrading to new releases or helping to test
release candidates.*

- [LND 0.11.0-beta][lnd 0.11.0-beta] is now released.   This new major
  version allows accepting [large channels][topic large channels] (by
  default, this is off) and contains numerous improvements to its
  backend features that may be of interest to advanced users (see the
  [release notes][lnd 0.11.0-beta]).

## Notable code and documentation changes

*Notable changes this week in [Bitcoin Core][bitcoin core repo],
[C-Lightning][c-lightning repo], [Eclair][eclair repo], [LND][lnd repo],
[Rust-Lightning][rust-lightning repo], [libsecp256k1][libsecp256k1 repo],
[Hardware Wallet Interface (HWI)][hwi], [Bitcoin Improvement Proposals
(BIPs)][bips repo], and [Lightning BOLTs][bolts repo].*

- [Bitcoin Core #14582][] and [#19743][Bitcoin Core #19743] wallet: always do partial spends FIXME:jonatack

- [Bitcoin Core #19550][] adds a new `getindexinfo` RPC that indicates
  which optional indexes have been enabled and how many blocks they've

>   indexed so far.

Reviewer: for each index, it indicates the sync status and blockheight

  so that they can each take a `locktime` parameter that specifies the
  nLockTime of the transaction to create.  The PR notes that this is
  "required for [dual funding][bolts #524], where the opener sets [the
  locktime]".

- [BIPs #955][] BIP174: add hash preimage fields to inputs FIXME:dongcarl

- [BOLTs #688][] adds support for [anchor outputs][topic anchor outputs]
  to the LN specification.  This extends commitment transactions with
  two extra outputs---one for each counterparty---which can be used for
  Child Pays For Parent ([CPFP][topic cpfp]) fee bumping.   With this
  change, it becomes possible to fee bump commitment transactions that
  may have been signed days or weeks earlier, when the current feerates
  would've been hard to predict.  In practice, LN nodes using anchor
  outputs should pay lower fees normally (because there's no longer any
  incentive to overestimate fees) and also have greater security
  (because, if feerates do increase beyond what was predicted, the node
  can fee bump its commitment transaction).  Various degrees of support
  for anchor outputs have already been merged into several LN

>   implementations monitored by Optech.

Reviewer 2: perhaps remove the "monitored by Optech"

Reviewer: I hesitated to comment on this as well, wondering "which impls does Optech not monitor?" while thinking that, knowing Dave, there must be a specific reason for writing this.

Author: We (I) monitor the ones listed in in the *notable code changes* section.  The reason for appending that text was as a lead-in for linking to our summaries of the PRs for anchor outputs from past newsletters; however, I got lazy and dropped that.  I'll drop this vestigial text too.

  transactions attempting theft).  In general, it appeared that all
  major concerns were addressed and that the protocol would provide
  several advantageous privacy benefits.

## Selected Q&A from Bitcoin StackExchange

*[Bitcoin StackExchange][bitcoin.se] is one of the first places Optech
contributors look for answers to their questions---or when we have a
few spare moments to help curious or confused users.  In
this monthly feature, we highlight some of the top-voted questions and
answers posted since our last update.*

{% comment %}<!-- https://bitcoin.stackexchange.com/search?tab=votes&q=created%3a1m..%20is%3aanswer -->{% endcomment %}
{% assign bse = "https://bitcoin.stackexchange.com/a/" %}

FIXME:bitschmidty

## Releases and release candidates

*New releases and release candidates for popular Bitcoin infrastructure
projects.  Please consider upgrading to new releases or helping to test
release candidates.*

- [LND 0.11.0-beta][lnd 0.11.0-beta] is now released.   This new major
  version allows accepting [large channels][topic large channels] (by
  default, this is off) and contains numerous improvements to its

>   backend features that may be of interest to advanced users (see the
>   [release notes][lnd 0.11.0-beta]).

Reviewer 3: Probably fine just to link once:
```suggestion
  backend features that may be of interest to advanced users.
```

## Notable code and documentation changes

*Notable changes this week in [Bitcoin Core][bitcoin core repo],
[C-Lightning][c-lightning repo], [Eclair][eclair repo], [LND][lnd repo],
[Rust-Lightning][rust-lightning repo], [libsecp256k1][libsecp256k1 repo],
[Hardware Wallet Interface (HWI)][hwi], [Bitcoin Improvement Proposals
(BIPs)][bips repo], and [Lightning BOLTs][bolts repo].*

- [Bitcoin Core #14582][] and [#19743][Bitcoin Core #19743] wallet: always do partial spends FIXME:jonatack

- [Bitcoin Core #19550][] adds a new `getindexinfo` RPC that lists each
  optional index that has been enabled, whether it is fully synced to

>   the tip of the block chain, and (in case it hasn't been fully synced)

Reviewer 3: I don't think parentheses quite fit here. The case of not being fully synced is equal to the fully synced case which is not in parentheses in the previous clause.

```suggestion
  the tip of the block chain, and if not fully synced,
```

  how many blocks have been indexed so far.

- [C-Lightning #3954][] updates both the `fundpsbt` and `utxopsbt` RPCs
  so that they can each take a `locktime` parameter that specifies the
  nLockTime of the transaction to create.  The PR notes that this is
  "required for [dual funding][bolts #524], where the opener sets [the

>   locktime]".

Reviewer 3: ```suggestion
  locktime]."
```

Author: What you propose is indeed the normal American style but a number of programmers have complained to me about "closing" the sentence before "closing" a quote that was "opened" within that sentence.  I did a quick search and it seems maybe other programing documentation writers have adopted the same policy: `https://github.com/MicrosoftDocs/microsoft-style-guide/issues/99`

So maybe this is something we discuss?  A quick grep says we already have many cases of `\."`, so I think this is ok to leave for now.  Does anyone have a strong opinion?  I have a slight preference towards the normal `."` style Jonas recommends.

Reviewer: +1 for `."`

- [BIPs #955][] BIP174: add hash preimage fields to inputs FIXME:dongcarl

- [BOLTs #688][] adds support for [anchor outputs][topic anchor outputs]
  to the LN specification.  This extends commitment transactions with
  two extra outputs---one for each counterparty---which can be used for
  Child Pays For Parent ([CPFP][topic cpfp]) fee bumping.   With this
  change, it becomes possible to fee bump commitment transactions that
  may have been signed days or weeks earlier, when the current feerates

>   would've been hard to predict.  In practice, LN nodes using anchor
>   outputs should pay lower fees normally (because there's no longer any
>   incentive to overestimate fees) and also have greater security
>   (because, if feerates do increase beyond what was predicted, the node
>   can fee bump its commitment transaction).  Various degrees of support

Reviewer 3: Suggest breaking into two sentences to match the two benefits of anchor outputs:
```suggestion
  would've been hard to predict.  In practice, LN nodes using anchor
  outputs should normally pay lower fees because there's no longer any
  incentive to overestimate fees. Anchor outputs also provide greater security
  because, if feerates do increase beyond what was predicted, the node
  can fee bump its commitment transaction.  Various degrees of support
```

---
title: 'Bitcoin Optech Newsletter #112'
permalink: /en/newsletters/2020/08/26/
name: 2020-08-26-newsletter
slug: 2020-08-26-newsletter
type: newsletter
layout: newsletter
lang: en
---
This week's newsletter links to a discussion about routed coinswaps and
includes our regular sections with summaries of questions and answers
from the Bitcoin StackExchange, releases and release candidates, and
notable changes to popular Bitcoin infrastructure software.

## Action items

*None this week.*

## News

- **Discussion about routed coinswaps:** Chris Belcher [posted][belcher
  coinswap] a design document for a routed multi-transaction coinswap
  implementation to the Bitcoin-Dev mailing list as a follow-up to his
  previous post in May (see [Newsletter #100][news100 coinswap]).
  Comments focused on checking that the
  protocol was safe both in its use of cryptography and in ensuring that

>   the expected transactions would get confirmed (rather than alternative

Reviewer: nit, s/get/be/

  transactions attempting theft).  Discussion was still ongoing at the
  time of this writing.

## Selected Q&A from Bitcoin StackExchange

*[Bitcoin StackExchange][bitcoin.se] is one of the first places Optech
contributors look for answers to their questions---or when we have a
few spare moments to help curious or confused users.  In
this monthly feature, we highlight some of the top-voted questions and
answers posted since our last update.*

{% comment %}<!-- https://bitcoin.stackexchange.com/search?tab=votes&q=created%3a1m..%20is%3aanswer -->{% endcomment %}
{% assign bse = "https://bitcoin.stackexchange.com/a/" %}

- [What is the motivation for separating Bitcoin Core into independent node, wallet, and GUI processes?]({{bse}}98398)
  Michael Folkson outlines the benefits of Russell Yanofsky's [efforts][bitcoin
  core process separation] to separate Bitcoin Core into different `bitcoin-node`,
  `bitcoin-wallet`, and `bitcoin-gui` processes. Modularity would allow users more
  flexibility in their setups and developers maintainability and security benefits.

- [What's the most efficient way to create a raw transaction with a specific fee rate?]({{bse}}98392)

>   User Darius asks about the interplay between UTXO selection and feerate when

Reviewer: "User Darius" is ambiguous. Maybe "User *Darius* ..." or "A user, Darius, ..."

  For this reason, `avoidpartialspends` is disabled by default unless
  the `avoid_reuse` flag is enabled for the wallet (see [Newsletter
  #52][news52 avoid_reuse]).  It is for this default case that maxapsfee
  was conceived---its addition gives users a choice between three
  configurations:

    1. `-maxapsfee=-1`: partial spend avoidance is completely
       disabled to optimize for faster fee calculations, which may be
       useful for very large wallets with many UTXOs.

    2. `-maxapsfee=0` (the default value): fee calculations are
       made using both coin selection algorithms.  Whichever result is
       cheaper is used; if they both result in the same cost, partial
       spend avoidance is used.

    3. `maxapsfee` set to greater than `0`: partial spend avoidance is
       used whenever the maximum additional cost it adds to the
       transaction is the passed amount.  For example,
       `-maxapsfee=0.00001000` means the wallet will avoid partial
       spends if the absolute fee difference is up to 1,000 sats.

- [Bitcoin Core #19550][] adds a new `getindexinfo` RPC that lists each
  optional index that has been enabled, how many blocks have been

>   indexed so far, and whether those blocks constitute syncing all the way

Reviewer: maybe "whether those blocks are synced to the tip of the block chain" (simpler)

  was [found to be necessary for miniscript-aware finalizers][psbt ext], though
  they can be used by any PSBT finalizer needing to satisfy hash preimage
  challenges (e.g. for onchain LN commitment transactions).

- [BOLTs #688][] adds support for [anchor outputs][topic anchor outputs]
  to the LN specification.  This extends commitment transactions with
  two extra outputs---one for each counterparty---which can be used for
  Child Pays For Parent ([CPFP][topic cpfp]) fee bumping.   With this
  change, it becomes possible to fee bump commitment transactions that
  may have been signed days or weeks earlier, when the current feerates
  would've been hard to predict.  In practice, LN nodes using anchor
  outputs should normally pay lower fees because there's no longer any
  incentive to overestimate fees. Anchor outputs also provide greater security
  because, if feerates do increase beyond what was predicted, the node
  can fee bump its commitment transaction.  Various degrees of support
  for anchor outputs have already been merged into several LN
  implementations.

- [BOLTS #785][] updates the minimum CLTV expiry delta to 18 blocks.  To

>   ensure that the latest state gets recorded onchain, a channel should

Reviewer: s/gets/is/

>   be unilaterally closed when there are only this many blocks until an

Reviewer: this = minimum expiry delta?

  LN payment has to be settled.  However, the higher the expiry is, the
  more time a payment could become temporarily stuck in an open channel
  (either by accident or deliberately).  This has led some LN
  implementations to use route-finding algorithms that optimize for
  routes with low CLTV expiry deltas, which has in turn led some users
  to set their deltas to values that are especially unsafe.  This new
  minimum should help prevent inexperienced users from naively setting

>   an unsafe value.

Reviewer: "unsafe" mentioned twice without saying why it is unsafe. A risk associated with higher expiry values is described above; the same for low values may be good here

Author: I think it is explained, "To ensure that the latest state is recorded onchain [before] an LN payment has to be settled."

- [What is the motivation for separating Bitcoin Core into independent node, wallet, and GUI processes?]({{bse}}98398)
  Michael Folkson outlines the benefits of Russell Yanofsky's [efforts][bitcoin
  core process separation] to separate Bitcoin Core into different `bitcoin-node`,
  `bitcoin-wallet`, and `bitcoin-gui` processes. Modularity would allow users more
  flexibility in their setups and developers maintainability and security benefits.

- [What's the most efficient way to create a raw transaction with a specific fee rate?]({{bse}}98392)
  User Darius asks about the interplay between UTXO selection and feerate when
  constructing a transaction. Murch provides an overview of the effective value
  approach to coin selection, consideration of [change avoidance][bitcoin wiki change avoidance],
  and attention to signature length.

- [Will there be a testnet4? Or do we not need a testnet reset once we have Signet?]({{bse}}98579)
  Michael Folkson explains the considerations involving a potential testnet
  reset or introduction of a new [signet][topic signet] network. A [related Stack
  Exchange question][stack exchange signet setup] this month also summarizes two

>   options for working with signet.

Reviewer: Probably out of scope, but a signet meta-issue was opened today by Michael as well and discussion has begun: https://github.com/bitcoin/bitcoin/issues/19787

(BIPs)][bips repo], and [Lightning BOLTs][bolts repo].*

- [Bitcoin Core #14582][] and [#19743][Bitcoin Core #19743] add a new
  `maxapsfee` ("max avoid partial spends fee") configuration option to
  specify the maximum amount of extra fee you're willing to pay to avoid
  partial spends when the existing `avoidpartialspends` configuration
  option is disabled.

  Enabling `avoidpartialspends` improves privacy by spending from
  addresses only once (see [Newsletter #6][news6 avoidpartial]), but it
  may result in slightly higher fees due to spending all inputs received
  to the same address when only a subset of those inputs might be needed.
  For this reason, `avoidpartialspends` is disabled by default unless
  the `avoid_reuse` flag is enabled for the wallet (see [Newsletter

>   #52][news52 avoid_reuse]).  The new `maxapsfee` feature now gives

Reviewer: The first paragraph is much better now.

The second paragraph perhaps lost some meaning and transition that was provided by "It is for this default case that `maxapsfee` was conceived", though I'll concede "conceived" may be a bit of an unusual word for the writing here. 

ISTM that "The new `maxapsfee` feature..." is an intro phrase that should be the start of a new paragraph.

I thought that "not more than" and "not greater than" seemed simpler and more concise than "less than or equal to" but deferring to your judgement.

Reviewer: Thanks for adding the topic links!

Author: @jonatack tried to address your concerns.

The problem I had with "**are not**" was that it always appeared in the same clause as "**are avoided"**.  Having two negatives in the same clause tends to make it hard to parse, at least for a attention deficit like me.  E.g., I found this challenging:

> partial spends are avoided if the fees are not higher than regular coin selection

However, if my edits aren't an obvious improvement, they're not good enough, so I tried again.  Let me know if you think this is better or if I should go back to your original.

  to the same address when only a subset of those inputs might be needed.
  For this reason, `avoidpartialspends` is disabled by default unless
  the `avoid_reuse` flag is enabled for the wallet (see [Newsletter

>   #52][news52 avoid_reuse]).  It is for this default case that maxapsfee

Reviewer: unsure if "maxapsfee" was intentionally left without code markdown in this line, but it's probably  fine like this as well

  was conceived---its addition gives users a choice between three
  configurations:

    1. `-maxapsfee=-1`: partial spend avoidance is completely
       disabled to optimize for faster fee calculations, which may be
       useful for very large wallets with many UTXOs.

    2. `-maxapsfee=0` (the default value): fee calculations are
       made using both coin selection algorithms.  Whichever result is
       cheaper is used; if they both result in the same cost, partial
       spend avoidance is used.

    3. `maxapsfee` set to greater than `0`: partial spend avoidance is
       used whenever the maximum additional cost it adds to the

>        transaction is the passed amount.  For example,

Reviewer: "whenever the maximum additional cost it adds to the transaction is the passed amount" seems less clear to me than the first and second versions, but I'll defer to any other opinion here

  absolute fee difference is not more than 1000 sats.

- [Bitcoin Core #19550][] adds a new `getindexinfo` RPC that lists each
  optional index that has been enabled, how many blocks have been

>   indexed so far, and whether those blocks constitue syncing all the way

Reviewer: constitue?

  "required for [dual funding][bolts #524], where the opener sets [the
  locktime]".

- [BIPs #955][] updates [BIP174][]to standardize supplying hash preimages in
  [PSBT][topic psbt] input records. The standardization of these preimage fields

>   were [found to be necessary for miniscript aware finalizers][psbt ext], though

Reviewer: s/were/was/ and missing hyphen in "miniscript aware"

- [BIPs #955][] updates [BIP174][] to standardize supplying hash preimages in
  [PSBT][topic psbt] input records. The standardization of these preimage fields

>   were [found to be necessary for miniscript-aware finalizers][psbt ext], though

Reviewer: I still think s/were/was/ here, as the subject is the (singular) standardization of the fields, not the (plural) fields.

Author: Sorry, I looked right past that in your original feedback.
